File: ocaml/bin/p0040.ml
(* Project Euler: Problem 40 *)

(*
    0.123456789 | 10111213...979899 | 100101102...997998999  | 100010011002  ...
      ---------   -----------------   ---------------------   -----------------
  len: 1 * 9       2 * 90              3 * 900                 4 * 9000      ...
       1 * 9 * 1   2 * 9 * 10          3 * 9 * 100             4 * 9 * 1000  ...
         --> block_num * 9 * base

  block #1: 1-digit number
  block #2: 2-digits number
  block #3: 3-digits number
    ...
  block #n: n-digits number
 *)

open Core

(*
    0.123456789 | 10111213...979899 | 100101102...997998999  | 100010011002  ...
                       ^[d15]=2
    example (d 15)
      pos = 15 > 1 * 9 * (10 ** 0) = 9
      pos <- 15 - 9 = 6

      pos = 6 <= 2 * 9 * (10 ** 1) = 180
      q <- (6 - 1) / 2 = 2, r <- (6 - 1) % 2 = 1
      num <- 10 ** (2 - 1) + q = 10 + 2 = 12
      d[15] = num / (10 ** (2 - r - 1)) % 10
            = 12 / (10 ** 0) % 10
            = 2
 *)
let d nth =
  let rec loop pos ndigits =
    let block_size = ndigits * 9 * (Int.pow 10 (ndigits - 1)) in
    if pos > block_size then
      loop (pos - block_size) (succ ndigits)
    else
      let q, r = (pos - 1) / ndigits, (pos - 1) mod ndigits in
      let num = (Int.pow 10 (ndigits - 1)) + q in
      (num / (Int.pow 10 (ndigits - r - 1))) mod 10
  in
  loop nth 1

let solve () =
  (d 1) * (d 10) * (d 100) * (d 1_000) * (d 10_000) * (d 100_000) * (d 1_000_000)

let exec () =
  Int.to_string (solve ())

let () = Euler.Task.run exec

File: rust/src/bin/p0040.rs
// Project Euler: Problem 40

euler::run_solver!(40);

fn solve() -> String {
    compute().to_string()
}

fn compute() -> u32 {
    [1, 10, 100, 1_000, 10_000, 100_000, 1_000_000]
        .into_iter()
        .map(d)
        .product()
}

fn d(mut pos: u32) -> u32 {
    let mut n_digits = 1;

    while pos > n_digits * 9 * (10_u32.pow(n_digits - 1)) {
        pos -= n_digits * 9 * (10_u32.pow(n_digits - 1));
        n_digits += 1;
    }
    let q = (pos - 1) / n_digits;
    let r = (pos - 1) % n_digits;
    let num = 10_u32.pow(n_digits - 1) + q;

    (num / (10_u32.pow(n_digits - r - 1))) % 10
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0040() {
        assert_eq!(compute(), 210);
    }
}
